pypeit/datamodel.py

```diff
@@ -0,0 +1,942 @@
+
+        def _validate(self):
+            if self.func not in ['add', 'sub']:
+                raise ValueError('Function must be either \'add\' or \'sub\'.')
+            # This is here because I don't want to overwrite something that
+            # might have been read in from an HDU, even though they should
+            # nominally be the same!
+            if self.out is None:
+                print('Assigning out!')
+                if self.func is None:
+                    raise ValueError('Do not know how to construct out attribute!')
+                self.out = self.inp1 + self.inp2 if self.func == 'add' else self.inp1 - self.inp2
+
+
+With this implementation:
+
+    - The instantiation of the ``BadInitContainer`` will fail because
+      the init arguments all need to be part of the datamodel for it
+      to work::
+
+        x = np.arange(10)
+        y = np.arange(10)+5
+
+        data = BadInitContainer(x,y)
+        # Fails with AttributeError
+
+    - The following instantiation is fine because
+      ``DubiousInitContainer`` handles the fact that some of the
+      arguments to :func:`__init__` are not part of the datamodel::
+
+        data = DubiousInitContainer(x,y)
+        print(np.array_equal(data.out, data.inp1+data.inp2))
+        # True
+
+    - One component of the data model wasn't instantiated, so it will
+      be None::
+        
+        print(data.alt is None)
+        # True
+
+    - The problem with ``DubiousInitContainer`` is that it has
+      attributes that cannot be reinstantiated from what's written to
+      disk. That is, :class:`DataContainers` aren't really fully
+      formed objects unless all of its relevant attributes are
+      components of the data model::
+
+        _data = DubiousInitContainer.from_hdu(hdu)
+        print(_data.func == DubiousInitContainer(x,y).func)
+        # False
+    
+    - This is solved by adding func to the datamodel::
+    
+        data = ComplexInitContainer(x,y)
+        _data = ComplexInitContainer.from_hdu(data.to_hdu(add_primary=True))
+        print(data.func == _data.func)
+        # True
+
+----
+
+.. include common links, assuming primary doc root is up one directory
+.. include:: ../links.rst
+"""
+import os
+import warnings
+
+from IPython import embed
+
+import numpy as np
+
+from astropy.io import fits
+from astropy.table import Table
+
+from pypeit import io
+
+class DataContainer:
+    """
+    Defines an abstract class for holding and manipulating data.
+
+    The primary utilities of the class are:
+        - Attributes can be accessed normally or as expected for a :obj:`dict`
+        - Attributes and items are restricted to conform to a specified data model.
+
+    This abstract class should only be used as a base class.
+
+    Derived classes must do the following:
+
+        - Define a datamodel
+        - Provide an :func:`__init__` method that defines the
```

Comment:
Somewhere, maybe here, we should make it clear that this object can have attributes that are *not* part of `datamodel`.  But that they need to be initialized in the `init()` or a method called from within `init()`.

Reply:
I added a note about this both in the top-level docstring and below this list.